cpp/src/copying/scatter.cu

```diff
@@ -34,20 +35,72 @@ using bit_mask::bit_mask_t;
 
 namespace cudf {
 namespace detail {
-  
-__global__ void invert_map(gdf_index_type gather_map[], const gdf_size_type destination_rows,
-                            gdf_index_type const scatter_map[], const gdf_size_type source_rows){
-  gdf_index_type source_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+template <typename map_type>
+__global__ void invert_map(map_type gather_map[], const gdf_size_type destination_rows,
+			   map_type const scatter_map[], const gdf_size_type source_rows){
+  map_type source_row = threadIdx.x + blockIdx.x * blockDim.x;
   if(source_row < source_rows){
-    gdf_index_type destination_row = scatter_map[source_row];
+    map_type destination_row = scatter_map[source_row];
     if(destination_row < destination_rows){
       gather_map[destination_row] = source_row;
     }
   }
 }
 
-void scatter(table const* source_table, gdf_index_type const scatter_map[],
-            table* destination_table) {
+struct invert_map_kernel_forwarder {
+template <typename map_type, std::enable_if_t<std::is_integral<map_type>::value>* = nullptr>
+void operator()(gdf_column gather_map,
+		const gdf_size_type num_destination_rows,
+		const gdf_column scatter_map,
+		const gdf_size_type num_source_rows) {
+
+  gdf_scalar fill_value;
+  fill_value.dtype = gdf_dtype_of<map_type>();
+  fill_value.is_valid = true;
+
+  switch (gdf_dtype_of<map_type>()) {
+  case GDF_INT8:
+    fill_value.data.si08 = -1;
+    break;
+  case GDF_INT16:
+    fill_value.data.si16 = -1;
+    break;
+  case GDF_INT32:
+    fill_value.data.si32 = -1;
+    break;
+  case GDF_INT64:
+    fill_value.data.si64 = -1;
+    break;
+  default:
+    CUDF_FAIL("Invalid scatter map type");
+  }
+
+  constexpr int block_size = 256;
+  const gdf_size_type invert_grid_size =
+    (num_destination_rows + block_size - 1) / block_size;
+
+  cudf::fill(&gather_map, fill_value, 0, gather_map.size);
```

Comment:
I think you could save the cost of the fill by building the initialized index filling into the `invert_map` kernel?

Reply:
Done